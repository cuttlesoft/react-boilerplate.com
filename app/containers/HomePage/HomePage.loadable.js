import React from 'react'
import loadable from 'utils/loadable'
import { LoadingIndicator } from 'components/LoadingIndicator'

/**
 * Asynchronously loads the component for HomePage
 */
export default loadable(() => import('./HomePage'), {
  fallback: <LoadingIndicator />,
})
