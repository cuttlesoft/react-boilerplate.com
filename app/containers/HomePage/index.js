/*
 * HomePage
 *
 * This is the first thing users see of our App, at the '/' route
 */

import React, { useContext, useEffect, useState, memo } from 'react'
import { Helmet } from 'react-helmet'
import { FormattedMessage } from 'react-intl'
import { observer } from 'mobx-react'

import { Header } from 'components/Header'
import ReposList from 'components/ReposList'
import AtPrefix from './AtPrefix'
import CenteredSection from './CenteredSection'
import Form from './Form'
import Input from './Input'
import Section from './Section'
import messages from './messages'

import { GitHubStoreContext } from '../../stores/GithubStore'

export const HomePage = observer(() => {
  const gitHubStore = useContext(GitHubStoreContext)
  const [searchTerm, setSearchTerm] = useState('')

  const onSubmitForm = evt => {
    if (evt !== undefined && evt.preventDefault) evt.preventDefault()
    gitHubStore.getRepos(searchTerm)
  }

  useEffect(() => {
    // When initial state username is not null, submit the form to load repos
    if (gitHubStore.currentUser && gitHubStore.currentUser.trim().length > 0)
      onSubmitForm(gitHubStore.currentUser)
  }, [])

  return (
    <article>
      <Helmet>
        <title>Home Page</title>
        <meta name="description" content="A React.js Boilerplate application homepage" />
      </Helmet>
      <div>
        <CenteredSection>
          <Header level="2">
            <FormattedMessage {...messages.startProjectHeader} />
          </Header>
          <Header level="2">
            <FormattedMessage {...messages.startProjectMessage} />
          </Header>
        </CenteredSection>
        <Section>
          <Header level="2">
            <FormattedMessage {...messages.trymeHeader} />
          </Header>
          <Form onSubmit={onSubmitForm}>
            <label htmlFor="searchTerm">
              <FormattedMessage {...messages.trymeMessage} />
              <AtPrefix>
                <FormattedMessage {...messages.trymeAtPrefix} />
              </AtPrefix>
              <Input
                id="searchTerm"
                type="text"
                placeholder="mxstbr"
                value={searchTerm}
                onChange={e => setSearchTerm(e.target.value)}
              />
            </label>
          </Form>
          <ReposList
            loading={gitHubStore.isLoading}
            error={gitHubStore.hasError}
            repos={gitHubStore.repos}
          />
        </Section>
      </div>
    </article>
  )
})

export default memo(HomePage)
