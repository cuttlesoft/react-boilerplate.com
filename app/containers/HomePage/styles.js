import styled from 'styled-components'

export const AtPrefix = styled.span`
  color: black;
  margin-left: 0.4em;
`

export const Form = styled.form`
  margin-bottom: 1em;
`

export const Input = styled.input`
  outline: none;
  border: none;
  border-bottom: 1px dotted #999;
  background-color: transparent;
`

export const Section = styled.section`
  margin: 3em auto;

  &:first-child {
    margin-top: 0;
  }
`
export const CenteredSection = styled(Section)`
  text-align: center;
`
