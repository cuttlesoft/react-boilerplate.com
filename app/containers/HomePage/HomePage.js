import React, { useContext, useEffect, useState, memo } from 'react'
import { Helmet } from 'react-helmet'
import { FormattedMessage } from 'react-intl'
import { observer } from 'mobx-react'

// Components
import { Header } from 'components/Header'
import { ReposList } from 'components/ReposList'

// Messages
import messages from './HomePage.messages'

// Styles
import { AtPrefix, CenteredSection, Form, Input, Section } from './HomePage.styles'

// Context / Stores
import { GitHubStoreContext } from '../../stores/GithubStore'

/*
 * HomePage
 *
 * The HomePage renders:
 * - The page title and metadata using Helmet
 * - The page header and description
 * - The search input and repository listing for a given user
 *
 * A user can:
 * - Show Github repositories by username
 *   - When a user inputs a username into the input field then hits the
 *     enter key, the form is submitted and search results are retrieved
 *   - Repositories are displayed using the ReposList component, which handles
 *     loading and error messages
 */
export const HomePage = observer(() => {
  const gitHubStore = useContext(GitHubStoreContext)
  const [searchTerm, setSearchTerm] = useState('')

  /**
   * Prevent the default behavior from occuring (for example, a page refresh),
   * then initiate fetching repositories based on the current `searchTerm`
   *
   * @param {object} evt The sythetic event from firing the function via
   *   `onSubmit`
   */
  const onSubmitForm = evt => {
    evt.preventDefault()
    gitHubStore.getRepos(searchTerm)
  }

  /**
   * When the initial username from the store is not null, set the local state
   * for `searchTerm` to the provided username and initiate fetching
   * repositories
   */
  useEffect(() => {
    if (gitHubStore.currentUser && gitHubStore.currentUser.trim().length > 0) {
      setSearchTerm(gitHubStore.currentUser)
      gitHubStore.getRepos(gitHubStore.currentUser)
    }
  }, [])

  return (
    <article>
      <Helmet>
        <title>Home Page</title>
        <meta name="description" content="A React.js Boilerplate application homepage" />
      </Helmet>

      <CenteredSection>
        <Header level="2" color="brand">
          <FormattedMessage {...messages.startProjectHeader} />
        </Header>
        <p>
          <FormattedMessage {...messages.startProjectMessage} />
        </p>
      </CenteredSection>

      <Section>
        <Header level="2">
          <FormattedMessage {...messages.trymeHeader} />
        </Header>

        <Form data-testid="search-form" onSubmit={onSubmitForm}>
          <label htmlFor="searchTerm">
            <FormattedMessage {...messages.trymeMessage} />

            <AtPrefix>
              <FormattedMessage {...messages.trymeAtPrefix} />
            </AtPrefix>

            <Input
              id="searchTerm"
              type="text"
              placeholder="mxstbr"
              value={searchTerm}
              onChange={e => setSearchTerm(e.target.value)}
            />
          </label>
        </Form>

        <ReposList
          loading={gitHubStore.isLoading}
          error={gitHubStore.hasError}
          repos={gitHubStore.repos}
        />
      </Section>
    </article>
  )
})

export default memo(HomePage)
