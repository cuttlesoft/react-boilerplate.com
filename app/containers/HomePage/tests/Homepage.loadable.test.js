/**
 * Test the HomePage
 */
import React from 'react'
import { waitForElement, render } from '@testing-library/react'
import HomePage from '../HomePage.loadable'

test('should lazy load and match the snapshot', async () => {
  const { container } = render(<HomePage />)
  const lazyComponent = await waitForElement(() => container)
  await expect(lazyComponent).toMatchSnapshot()
})
