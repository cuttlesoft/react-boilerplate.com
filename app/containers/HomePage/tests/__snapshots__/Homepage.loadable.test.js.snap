// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should lazy load and match the snapshot 1`] = `
<div>
  <article>
    <div>
      <section
        class="styles__Section-w51kdb-3 styles__CenteredSection-w51kdb-4 hZTVSk"
      >
        <h2
          class="StyledHeading-sc-1rdh4aw-0 yhirM"
        >
          <span>
            Start your next react project in seconds
          </span>
        </h2>
        <p>
          <span>
            A highly scalable, offline-first foundation with the best DX and a focus on performance and best practices
          </span>
        </p>
      </section>
      <section
        class="styles__Section-w51kdb-3 ktSOjr"
      >
        <h2
          class="StyledHeading-sc-1rdh4aw-0 yhirM"
        >
          <span>
            Try me!
          </span>
        </h2>
        <form
          class="styles__Form-w51kdb-1 hMNxRg"
        >
          <label
            for="searchTerm"
          >
            <span>
              Show Github repositories by
            </span>
            <span
              class="styles__AtPrefix-w51kdb-0 cRLVte"
            >
              <span>
                @
              </span>
            </span>
            <input
              class="styles__Input-w51kdb-2 hEplru"
              id="searchTerm"
              placeholder="mxstbr"
              type="text"
              value=""
            />
          </label>
        </form>
        <div
          class="Wrapper-sc-1umgotm-0 laCwDT"
        >
          <ul
            class="Ul-mwnq6h-0 jcxXFy"
          />
        </div>
      </section>
    </div>
  </article>
</div>
`;

exports[`should render and match the snapshot 1`] = `
<div>
  <article>
    <div>
      <section
        class="styles__Section-w51kdb-3 styles__CenteredSection-w51kdb-4 hZTVSk"
      >
        <h2
          class="StyledHeading-sc-1rdh4aw-0 yhirM"
        >
          <span>
            Start your next react project in seconds
          </span>
        </h2>
        <p>
          <span>
            A highly scalable, offline-first foundation with the best DX and a focus on performance and best practices
          </span>
        </p>
      </section>
      <section
        class="styles__Section-w51kdb-3 ktSOjr"
      >
        <h2
          class="StyledHeading-sc-1rdh4aw-0 yhirM"
        >
          <span>
            Try me!
          </span>
        </h2>
        <form
          class="styles__Form-w51kdb-1 hMNxRg"
        >
          <label
            for="searchTerm"
          >
            <span>
              Show Github repositories by
            </span>
            <span
              class="styles__AtPrefix-w51kdb-0 cRLVte"
            >
              <span>
                @
              </span>
            </span>
            <input
              class="styles__Input-w51kdb-2 hEplru"
              id="searchTerm"
              placeholder="mxstbr"
              type="text"
              value=""
            />
          </label>
        </form>
        <div
          class="Wrapper-sc-1umgotm-0 laCwDT"
        >
          <ul
            class="Ul-mwnq6h-0 jcxXFy"
          />
        </div>
      </section>
    </div>
  </article>
</div>
`;
